feat(components): [date-picker] add type yearrange for DatePicker

CherishTheYouth opened the pull request:
Please make sure these boxes are checked before submitting your PR, thank you!

- [X] Make sure you follow contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
- [X] Make sure you are merging your commits to `dev` branch.
- [X] Add some descriptions and refer to relative issues for your PR.

## Description

Add type `yearrange` for date-picker

## Related Issue

close #6388, close #2487, close #7051, close #2042

## Explanation of Changes
I implement this feature based on the `monthrange` and `daterange` date-picker with typescript.

Commit:
fix(components): fix bug #14637

CherishTheYouth committed:
Merge branch 'element-plus:dev' into dev

Commit:
refactor(components): [el-button]update props autoBlur name

Commit:
docs(components): [el-button] update docs ablout props clickAutoBlur

CherishTheYouth committed:
Merge branch 'dev' into dev

Commit:
fix(components): [date-picker]  fix bug  #14728

Commit:
feat(components): [date-picker] add yearrange

Commit:
feat(components): [date-picker] update yearrange codes

Commit:
feat(components): [date-picker] merge yearrange code to new branch

Commit:
feat(components): [date-picker]fix select range ui bug

Commit:
refactor(components): [date-picker]convert setup() to setup, add TS type

Commit:
refactor(components): refactor pane-year-range

Commit:
refactor(components): [date-picker] refactor panel-year-range styles

Commit:
docs(components): [date-picker] update year-range docs #6388 #2487 #7051

Commit:
style(components): [date-picker] remove repeat css

CherishTheYouth committed:
Merge branch 'dev' into feat/yearrange

FrontEndDog commented on docs/examples/button/click-auto-blur.vue:
remove this file.

FrontEndDog reviewed (changes requested):
Thanks for your contribution! It would be even better if you could add a test case!

CherishTheYouth commented:
> Thanks for your contribution! It would be even better if you could add a test case!

I will make time to add test case, thank you!

lzw-leaf commented:
╯︿╰，when will this commit be ready for production?

CherishTheYouth commented:
> ╯︿╰，when will this commit be ready for production?

抱歉哈，我还以为没人关注这个feature呢，我尽快弄下

CherishTheYouth committed:
Merge branch 'element-plus:dev' into dev

CherishTheYouth committed:
Merge branch 'dev' into feat/yearrange

upthen committed:
feat(components): [date-picker]add yearrange unit test

upthen committed:
feat: sync changes

Merge branch 'feat/yearrange' of https://github.com/CherishTheYouth/element-plus-yearrange into feat/yearrange

upthen committed:
fix(components): [date-picker] fix  unit test case failed error

CherishTheYouth commented:
> ╯︿╰，when will this commit be ready for production?

抱歉哈，我还以为没人关注这个feature呢，我尽快弄下了

> Thanks for your contribution! It would be even better if you could add a test case!


I have add some test cases just now, and it passed.

CherishTheYouth commented:
@FrontEndDog @btea Would you mind reviewing this pr in your free time, much appreciate. 感激！

CherishTheYouth committed:
Merge branch 'dev' into feat/yearrange

hibichann commented:
i really need this feature, btw there is a bug when change type from year/years to *range at shortcuts value callback(sadly, the best choice is yearrange, but i cant), im going to issue it at this weekend

btea commented on packages/components/date-picker/src/date-picker-com/basic-year-table.vue:
```diff
@@ -4,29 +4,26 @@
+          @keydown.enter.prevent.stop="handleYearTableClick"
+        >
+          <div>
+            <a class="cell">{{ cell.text }}</a>
```
This is not a link, and should be used as before with `span`.

btea commented on docs/examples/button/click-auto-blur.vue:
```diff
@@ -0,0 +1,15 @@
+
+<script setup lang="ts">
+import { Delete, Edit, Share } from '@element-plus/icons-vue'
+</script>
```
Why add an example to `button`? 🤔

CherishTheYouth replied:
Have been removed

btea commented on packages/components/date-picker/src/date-picker-com/panel-year-range.vue:
```diff
@@ -0,0 +1,263 @@
+}
+const getDefaultValue = () => {
+  let start: Dayjs
+  if (Array.isArray(defaultValue.value)) {
```
Use encapsulated methods.
```suggestion
  if (isArray(defaultValue.value)) {
```

CherishTheYouth replied:
Have import isArray from @element-plus/utils

btea commented on packages/components/date-picker/src/date-picker-com/basic-year-table.vue:
```diff
@@ -59,54 +67,182 @@ const startYear = computed(() => {
-const getCellKls = (year: number) => {
-  const kls: Record<string, boolean> = {}
+const getCellStyle = (cell: YearCell) => {
+  const style = {} as any
```
The types here should not be changed.

upthen committed:
fix(components): [date-picker] resolve some pr problems

upthen committed:
feat: sync changes from upstream.

Merge branch 'feat/yearrange' of https://github.com/CherishTheYouth/element-plus-yearrange into feat/yearrange

ZhangYC193 commented:
> i really need this feature, BTW there is a bug when change type from year/years to *range at shortcuts value callback(sadly, the best choice for me is yearrange, but i cant), im going to issue it at this weekend

me too

CherishTheYouth commented:
> > i really need this feature, BTW there is a bug when change type from year/years to *range at shortcuts value callback(sadly, the best choice for me is yearrange, but i cant), im going to issue it at this weekend
> 
> me too

我有基于 element-plus 写的本地版本，你要吗，这个pr不知道什么时候能合入，最早估计也要等下一个小版本了

hibichann commented:
> > > i really need this feature, BTW there is a bug when change type from year/years to *range at shortcuts value callback(sadly, the best choice for me is yearrange, but i cant), im going to issue it at this weekend
> > 
> > 
> > me too
> 
> 我有基于 element-plus 写的本地版本，你要吗，这个pr不知道什么时候能合入，最早估计也要等下一个小版本了

不太好，还是得从仓库下

ZhangYC193 commented:
> > > i really need this feature, BTW there is a bug when change type from year/years to *range at shortcuts value callback(sadly, the best choice for me is yearrange, but i cant), im going to issue it at this weekend
> > 
> > 
> > me too
> 
> 我有基于 element-plus 写的本地版本，你要吗，这个pr不知道什么时候能合入，最早估计也要等下一个小版本了

大佬，下个月能上吗

btea commented:
There are code conflicts that need to be resolved.

btea commented on packages/components/date-picker/src/date-picker-com/panel-year-range.vue:
```diff
@@ -0,0 +1,264 @@
+      {
+        'has-sidebar': Boolean($slots.sidebar) || hasShortcuts,
+      },
+    ]"
```
class has multiple values ​​that can be extracted into the following variables.

btea commented on packages/theme-chalk/src/date-picker/year-table.scss:
```diff
@@ -13,7 +13,7 @@
     width: 68px;
     text-align: center;
-    padding: 8px 0px;
+    padding: 20px 0px;
```
Why change the style here?

CherishTheYouth replied:
> Why change the style here?

Oh , not necessary indeed, have been reverted

tolking commented on packages/components/date-picker/src/props/basic-year-table.ts:
```diff
@@ -3,13 +3,9 @@ import { datePickerSharedProps, selectionModeWithDefault } from './shared'
+  ...datePickerSharedProps,
   selectionMode: selectionModeWithDefault('year'),
-} as const)
+})
```
```suggestion
} as const)
```

tolking commented:
@CherishTheYouth Do you have time to deal with relevant conflicts?

CherishTheYouth commented:
> @CherishTheYouth Do you have time to deal with relevant conflicts?

Sorry, a  little busy recently, I will resolve them soon, maybe this weekend.

upthen committed:
feat: Merge branch 'dev' into feat/yearrange

CherishTheYouth committed:
Merge branch 'element-plus:dev' into dev

upthen committed:
fix: merge form dev and resolve conflicts

upthen committed:
docs(components): [yearrange] update docs

upthen committed:
fix(docs): [yearrange] update example of year-range

upthen committed:
refactor(components): [date-picker] resolve some code review suggestions

CherishTheYouth commented:
> @CherishTheYouth Do you have time to deal with relevant conflicts?

It has been done, you could review again, thank you!

CherishTheYouth commented on packages/components/date-picker/src/composables/use-year-range-header.ts:
```diff
@@ -0,0 +1,76 @@
+
+  const leftLabel = computed(() => {
+    const leftStartDate = dayjs()
+      .year(Math.floor(leftDate.value.year() / 10) * 10)
```
Yes, it will

tolking commented on packages/components/date-picker/src/date-picker-com/basic-year-table.vue:
```diff
@@ -59,54 +67,165 @@ const startYear = computed(() => {
+  const selectedYear = row * 4 + column + startYear.value
+  const newDate = dayjs().year(selectedYear)
+
+  if (props.selectionMode === 'range') {
```
The code related to `if (props.selectionMode === 'years') {` should not be deleted here

tolking commented on packages/components/date-picker/src/date-picker-com/panel-date-pick.vue:
```diff
@@ -803,6 +805,13 @@ watch(
   { immediate: true }
 )
 
+// When close time-pick-panel, blur the timePickerInput
+watch(timePickerVisible, (newValue) => {
+  if (!newValue) {
+    timePickerInputRef.value?.blur()
+  }
```
It seems that there is no need to modify this file

tolking commented on packages/components/date-picker/src/date-picker-com/panel-month-range.vue:
```diff
@@ -176,10 +176,6 @@ type RangePickValue = {
-  // const minDate_ = modifyWithTimeString(val.minDate, defaultTime[0])
-  // const maxDate_ = modifyWithTimeString(val.maxDate, defaultTime[1])
-  // todo
   const minDate_ = val.minDate
```
It seems that there is no need to modify this file

tolking reviewed (commented):
I noticed that you may have made some mistakes when merging conflicts.

CherishTheYouth commented on packages/components/date-picker/src/composables/use-year-range-header.ts:
```diff
@@ -0,0 +1,76 @@
+  const leftLabel = computed(() => {
+    const leftStartDate = dayjs()
+      .year(Math.floor(leftDate.value.year() / 10) * 10)
+      .year()
```
Oh, you are right, it seems not

tolking replied:
Are there any differences between using both types of code in this file?

L38, L46 / L52, L57

upthen committed:
fix(components): [date-picker] resolve some CR problems

kooriookami commented:
![QQ_1722567942376](https://github.com/user-attachments/assets/51e69ee8-c2c2-4415-8b11-2b14b10e69d7)

Seems error occured.

upthen committed:
refactor(components): [date-picker] resolve some CR conversations

CherishTheYouth commented:
> ![QQ_1722568038467](https://private-user-images.githubusercontent.com/38392315/354446262-24e5ab9e-d226-4d35-b3b1-cf607cce119b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI2ODUwNjgsIm5iZiI6MTcyMjY4NDc2OCwicGF0aCI6Ii8zODM5MjMxNS8zNTQ0NDYyNjItMjRlNWFiOWUtZDIyNi00ZDM1LWIzYjEtY2Y2MDdjY2UxMTliLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODAzVDExMzI0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTgwMjBhNjBhZGU3OGExZWEyMmM2Mjc5MDczOTMyYWZhNjBiODRmNmYzOTM4MDUyZGU1MTIwYTZmZWZlMWE4MjcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Lg35NZW6vxH-ozfQ-V8j_jWJAZWuB2tFElLEoC6gg4E)
> 
> ![QQ_1722567942376](https://private-user-images.githubusercontent.com/38392315/354445973-51e69ee8-c2c2-4415-8b11-2b14b10e69d7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI2ODUwNjgsIm5iZiI6MTcyMjY4NDc2OCwicGF0aCI6Ii8zODM5MjMxNS8zNTQ0NDU5NzMtNTFlNjllZTgtYzJjMi00NDE1LThiMTEtMmIxNGIxMGU2OWQ3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODAzVDExMzI0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFjYzQyYzc5NTMxOTcwY2I4M2YzNWMxNWIyOGYzY2ZiNzhmYjk0MjA0NWFkODRiMWJhOGU3MDcwNGU3YjBkM2QmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.lRs1kEQ3xqFLmZDa7x31uT1f_CGJOftk0femph4PKCw)
> 
> Seems error occured.

Have resolved

CherishTheYouth committed:
Merge branch 'element-plus:dev' into dev

CherishTheYouth committed:
Merge branch 'element-plus:dev' into feat/yearrange

CherishTheYouth commented:
> ![QQ_1722568038467](https://private-user-images.githubusercontent.com/38392315/354446262-24e5ab9e-d226-4d35-b3b1-cf607cce119b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI2ODUwNjgsIm5iZiI6MTcyMjY4NDc2OCwicGF0aCI6Ii8zODM5MjMxNS8zNTQ0NDYyNjItMjRlNWFiOWUtZDIyNi00ZDM1LWIzYjEtY2Y2MDdjY2UxMTliLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODAzVDExMzI0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTgwMjBhNjBhZGU3OGExZWEyMmM2Mjc5MDczOTMyYWZhNjBiODRmNmYzOTM4MDUyZGU1MTIwYTZmZWZlMWE4MjcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Lg35NZW6vxH-ozfQ-V8j_jWJAZWuB2tFElLEoC6gg4E)
> 
> ![QQ_1722567942376](https://private-user-images.githubusercontent.com/38392315/354445973-51e69ee8-c2c2-4415-8b11-2b14b10e69d7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI2ODUwNjgsIm5iZiI6MTcyMjY4NDc2OCwicGF0aCI6Ii8zODM5MjMxNS8zNTQ0NDU5NzMtNTFlNjllZTgtYzJjMi00NDE1LThiMTEtMmIxNGIxMGU2OWQ3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODAzVDExMzI0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFjYzQyYzc5NTMxOTcwY2I4M2YzNWMxNWIyOGYzY2ZiNzhmYjk0MjA0NWFkODRiMWJhOGU3MDcwNGU3YjBkM2QmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.lRs1kEQ3xqFLmZDa7x31uT1f_CGJOftk0femph4PKCw)
> 
> Seems error occured.

When I run vitepress dev , I found this import not found, and it not found in dev too.
![image](https://github.com/user-attachments/assets/b08cf272-92e8-42b0-9142-73e359904ad8)

Could you have a look?

btea commented:
Run this script to generate the corresponding content.
https://github.com/element-plus/element-plus/blob/d94aa4cdb5a13ebc9330aefa8aaa4f2b5dcd87c4/docs/package.json#L8

CherishTheYouth commented:
> ![QQ_1722568038467](https://private-user-images.githubusercontent.com/38392315/354446262-24e5ab9e-d226-4d35-b3b1-cf607cce119b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI2ODcyNDgsIm5iZiI6MTcyMjY4Njk0OCwicGF0aCI6Ii8zODM5MjMxNS8zNTQ0NDYyNjItMjRlNWFiOWUtZDIyNi00ZDM1LWIzYjEtY2Y2MDdjY2UxMTliLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODAzVDEyMDkwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRmMWU2ODM2Nzg2ZDk2MjhhNzYwYzcxMWZiMGMwYjRmNTcwZjRlNGQyMzUyZTAxNGE0MjcyY2YzMDUwNWQyZmQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.yKYdS3I58dx6qoyCsQfPSwqqWzYsfBAm-usXZzIbpCM)
> 
> ![QQ_1722567942376](https://private-user-images.githubusercontent.com/38392315/354445973-51e69ee8-c2c2-4415-8b11-2b14b10e69d7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI2ODcyNDgsIm5iZiI6MTcyMjY4Njk0OCwicGF0aCI6Ii8zODM5MjMxNS8zNTQ0NDU5NzMtNTFlNjllZTgtYzJjMi00NDE1LThiMTEtMmIxNGIxMGU2OWQ3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODAzVDEyMDkwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNiZjhhYmZhMDdiZDk0YWQyZGI4NzI4M2Y5Yjc2M2YzOTZlMDFmYWUxOWNhMGYzNGE2MmU4ZDY2ODk5M2FiZWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.cmPfnDewzOtXU_K8jtpCe6WmkttjRTV3bsbBFGTykWY)
> 

![image](https://github.com/user-attachments/assets/f682ae5d-35dd-41e6-b771-4929cd1f107a)
Have tested it on my local, It seems to be OK

tolking commented on packages/components/date-picker/src/date-picker-com/panel-year-range.vue:
```diff
@@ -0,0 +1,264 @@
+  return [start, start.add(10, 'year')]
+}
+
+emit('set-picker-option', ['formatToString', formatToString])
```
We need to refer to the `panel-date-range.vue` to improve more methods.

```ts
emit('set-picker-option', ['isValidValue', isValidValue])
emit('set-picker-option', ['parseUserInput', parseUserInput])
emit('set-picker-option', ['formatToString', formatToString])
emit('set-picker-option', ['handleClear', handleClear])
```

Currently, an error will be reported after manually entering the time in the input box

upthen committed:
refactor(components): [date-picker] uniform date compute logic

upthen committed:
fix(components): [date-picker] error occurs when mannul input the date

btea commented:
There is a conflict between the code and the default branch, which needs to be resolved.

CherishTheYouth committed:
Merge branch 'element-plus:dev' into dev

upthen committed:
fix(components): [date-picker] resolve conflicts

CherishTheYouth commented:
> There is a conflict between the code and the default branch, which needs to be resolved.

Have resolved

btea commented on packages/components/date-picker/src/date-picker-com/panel-date-pick.vue:
```diff
@@ -249,7 +249,6 @@ const { shortcuts, disabledDate, cellClassName, defaultTime } = pickerBase.props
 
 const currentViewRef = ref<{ focus: () => void }>()
-
 const innerDate = ref(dayjs().locale(lang.value))
```
`panel-date-pick.vue` and `panel-month-range.vue` are not involved in this function and do not need to be modified.

upthen replied:
Have Reverted

btea replied:
It seems that no new commits are seen.

btea commented on packages/components/date-picker/src/date-picker-com/basic-year-table.vue:
```diff
@@ -4,29 +4,26 @@
+      <tr v-for="(row, key) in rows" :key="key">
+        <td
+          v-for="(cell, key_) in row"
+          :key="key_"
```
The current writing method will repeat the key.
```suggestion
          :key="key + '_' + key_"
```

btea commented on packages/components/date-picker/src/date-picker-com/basic-year-table.vue:
```diff
@@ -59,54 +67,181 @@ const startYear = computed(() => {
+    const column = target.cellIndex
+    const row = (target.parentNode as HTMLTableRowElement).rowIndex
+
+    const selectedYear = row * 4 + column + startYear.value
```
To get the year value of the currently clicked cell, it should be possible to directly use the original method to get the text content of the `td` tag. Are there any considerations for using the current process?

upthen replied:
IMO, this calculation will be more accurate,
First, dom can be changed manually, and if it is changed, the wrong value will be obtained. Second, if td needs to be written in a non-numeric form in some possible new feature, such as 二零二零, it will also be problematic to obtain td directly (although this requirement is unlikely to occur).
What do you think?

btea replied:
Make sense.

btea commented on packages/components/date-picker/src/date-picker-com/panel-year-range.vue:
```diff
@@ -0,0 +1,297 @@
+                ppNs.e('icon-btn'),
+                { [ppNs.is('disabled')]: !enableYearArrow },
+              ]"
+              class="d-arrow-right"
```
Multiple classes can be put together.

upthen committed:
refactor(components): [date-picker] resolve some CR conversations

tolking commented on packages/components/date-picker/src/date-picker-com/basic-year-table.vue:
```diff
@@ -4,29 +4,26 @@
+          :ref="(el) => isSelectedCell(cell) && (currentCellRef = el as HTMLElement)"
+          class="available"
+          :class="getCellKls(cell)"
+          :aria-selected="`${isSelectedCell(cell)}`"
```
```suggestion
          :aria-selected="isSelectedCell(cell)"
```

tolking reviewed (commented):
看起来有些问题并没有解决，我建议你查看下每条建议。另外避免改动不必要的文件

upthen committed:
refactor(components): [date-picker] resolve some CR conversations

CherishTheYouth commented:
> 看起来有些问题并没有解决，我建议你查看下每条建议。另外避免改动不必要的文件

好的， 昨天是我忘记推代码了，以后会注意，只修改相关文件

CherishTheYouth committed:
Merge branch 'dev' into feat/yearrange

kooriookami commented:
![QQ_1722906534693](https://github.com/user-attachments/assets/14c8704d-fd53-43f1-bff4-bea305116080)

![QQ_1722906511236](https://github.com/user-attachments/assets/4294ce1e-e15b-428f-a4b6-dcf45ec687c8)

![QQ_1722906709222](https://github.com/user-attachments/assets/c12cb097-c749-4e40-b761-69221cd85fc3)

![QQ_1722906687067](https://github.com/user-attachments/assets/25cdc8a5-a54e-4de8-a639-b1a9e4bb9322)

These two borders and widths are different. The month range has no problem, please take a look.

tolking commented on packages/components/date-picker/src/date-picker-com/basic-year-table.vue:
```diff
@@ -59,54 +67,180 @@ const startYear = computed(() => {
+  if (!target || !target.textContent || hasClass(target, 'disabled')) return
+
+  if (props.selectionMode === 'range') {
+    const column = target.cellIndex
+    const row = (target.parentNode as HTMLTableRowElement).rowIndex
+
+    const selectedYear = row * 4 + column + startYear.value
+    const newDate = dayjs().year(selectedYear)
+    if (!props.rangeState.selecting) {
+      emit('pick', { minDate: newDate, maxDate: null })
+      emit('select', true)
     } else {
-      emit('pick', Number(year))
+      if (props.minDate && newDate >= props.minDate) {
+        emit('pick', { minDate: props.minDate, maxDate: newDate })
+      } else {
+        emit('pick', { minDate: newDate, maxDate: props.minDate })
+      }
+      emit('select', false)
     }
+    return
+  }
+
+  const year = target.textContent || target.innerText
+  if (props.selectionMode === 'years') {
+    if (event.type === 'keydown') {
+      emit('pick', castArray(props.parsedValue), false)
+      return
+    }
+    const newValue = hasClass(target, 'current')
+      ? castArray(props.parsedValue).filter((d) => d?.year() !== Number(year))
+      : castArray(props.parsedValue).concat([dayjs(year)])
+
+    emit('pick', newValue)
+  } else {
+    emit('pick', Number(year))
+  }
```
How about

```ts
if (props.selectionMode === 'range') {

}

if (props.selectionMode === 'years') {

} else {

}
```

=>

```ts
if (props.selectionMode === 'range') {

} else if (props.selectionMode === 'years') {

} else {

}
```

upthen committed:
refactor(components): [date-picker] resolve some CR conversations

upthen committed:
Merge branch 'feat/yearrange' of https://github.com/CherishTheYouth/element-plus-yearrange into feat/yearrange

CherishTheYouth commented:
> ![QQ_1722906534693](https://private-user-images.githubusercontent.com/38392315/355278569-14c8704d-fd53-43f1-bff4-bea305116080.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI5NTM0NjQsIm5iZiI6MTcyMjk1MzE2NCwicGF0aCI6Ii8zODM5MjMxNS8zNTUyNzg1NjktMTRjODcwNGQtZmQ1My00M2YxLWJmZjQtYmVhMzA1MTE2MDgwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODA2VDE0MDYwNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTEyNTdiN2IyYzg4MGQ0NGRiYzUxZDI2OWU0ZjMwMTE5NjRjYjJhMWI5NjA5OTNlYjA4ZmNkNzAzODE5ZDcxMTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.RCzfCb3vDUrg4E_zzsecR9WfB6nzBFqg08Q6Hn9BTC4)
> 
> ![QQ_1722906511236](https://private-user-images.githubusercontent.com/38392315/355278521-4294ce1e-e15b-428f-a4b6-dcf45ec687c8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI5NTM0NjQsIm5iZiI6MTcyMjk1MzE2NCwicGF0aCI6Ii8zODM5MjMxNS8zNTUyNzg1MjEtNDI5NGNlMWUtZTE1Yi00MjhmLWE0YjYtZGNmNDVlYzY4N2M4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODA2VDE0MDYwNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNlNDI0YWFkYmU4YTRkNWRkNGQ5Mjg5NTQyMDVlNjMwMGQ0ZGEzM2M0ZmE2MzM2M2M1YjVjZjEwMTI4MDlmODEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.qOKEw7ocsLoEDntzzlftgm5n5kAu5xFLT6jy1fXFsDc)
> 
> ![QQ_1722906709222](https://private-user-images.githubusercontent.com/38392315/355278957-c12cb097-c749-4e40-b761-69221cd85fc3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI5NTM0NjQsIm5iZiI6MTcyMjk1MzE2NCwicGF0aCI6Ii8zODM5MjMxNS8zNTUyNzg5NTctYzEyY2IwOTctYzc0OS00ZTQwLWI3NjEtNjkyMjFjZDg1ZmMzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODA2VDE0MDYwNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg3ZmI3MmFjNzM2YTJlZjBhN2JmZjU4NTdlZGMxNjczMzVkMmUxNDRhZmMwMDUxODIxZDIyZjg3N2U0ZjA2Y2QmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.V4simG8rPmpOL72jWt2-eLaAok8TnNMKesitmtknbx8)
> 
> ![QQ_1722906687067](https://private-user-images.githubusercontent.com/38392315/355278904-25cdc8a5-a54e-4de8-a639-b1a9e4bb9322.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI5NTM0NjQsIm5iZiI6MTcyMjk1MzE2NCwicGF0aCI6Ii8zODM5MjMxNS8zNTUyNzg5MDQtMjVjZGM4YTUtYTU0ZS00ZGU4LWE2MzktYjFhOWU0YmI5MzIyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODA2VDE0MDYwNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThlY2M5MjYxNmE2MjRkMjQxMTgxNGQ5ZTc4ZTUxMWJlMDRjMjg1Mzk0MzFkMzNmNjZjMWU0MDMzZDE0NGEzYzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.9a-DETpicUQVAl4h9urHnDEaiZNZg7eQIjWmteuCidY)
> 
> Borders and widths of `Default` and `With quick options` are different. The month range has no problem, please take a look.

Something wrong with the panelKls, I have fixed it.

CherishTheYouth commented:
> > ![QQ_1722906534693](https://private-user-images.githubusercontent.com/38392315/355278569-14c8704d-fd53-43f1-bff4-bea305116080.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI5NTM0NjQsIm5iZiI6MTcyMjk1MzE2NCwicGF0aCI6Ii8zODM5MjMxNS8zNTUyNzg1NjktMTRjODcwNGQtZmQ1My00M2YxLWJmZjQtYmVhMzA1MTE2MDgwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODA2VDE0MDYwNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTEyNTdiN2IyYzg4MGQ0NGRiYzUxZDI2OWU0ZjMwMTE5NjRjYjJhMWI5NjA5OTNlYjA4ZmNkNzAzODE5ZDcxMTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.RCzfCb3vDUrg4E_zzsecR9WfB6nzBFqg08Q6Hn9BTC4)
> > ![QQ_1722906511236](https://private-user-images.githubusercontent.com/38392315/355278521-4294ce1e-e15b-428f-a4b6-dcf45ec687c8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI5NTM0NjQsIm5iZiI6MTcyMjk1MzE2NCwicGF0aCI6Ii8zODM5MjMxNS8zNTUyNzg1MjEtNDI5NGNlMWUtZTE1Yi00MjhmLWE0YjYtZGNmNDVlYzY4N2M4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODA2VDE0MDYwNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNlNDI0YWFkYmU4YTRkNWRkNGQ5Mjg5NTQyMDVlNjMwMGQ0ZGEzM2M0ZmE2MzM2M2M1YjVjZjEwMTI4MDlmODEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.qOKEw7ocsLoEDntzzlftgm5n5kAu5xFLT6jy1fXFsDc)
> > ![QQ_1722906709222](https://private-user-images.githubusercontent.com/38392315/355278957-c12cb097-c749-4e40-b761-69221cd85fc3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI5NTM0NjQsIm5iZiI6MTcyMjk1MzE2NCwicGF0aCI6Ii8zODM5MjMxNS8zNTUyNzg5NTctYzEyY2IwOTctYzc0OS00ZTQwLWI3NjEtNjkyMjFjZDg1ZmMzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODA2VDE0MDYwNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg3ZmI3MmFjNzM2YTJlZjBhN2JmZjU4NTdlZGMxNjczMzVkMmUxNDRhZmMwMDUxODIxZDIyZjg3N2U0ZjA2Y2QmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.V4simG8rPmpOL72jWt2-eLaAok8TnNMKesitmtknbx8)
> > ![QQ_1722906687067](https://private-user-images.githubusercontent.com/38392315/355278904-25cdc8a5-a54e-4de8-a639-b1a9e4bb9322.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI5NTM0NjQsIm5iZiI6MTcyMjk1MzE2NCwicGF0aCI6Ii8zODM5MjMxNS8zNTUyNzg5MDQtMjVjZGM4YTUtYTU0ZS00ZGU4LWE2MzktYjFhOWU0YmI5MzIyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODA2VDE0MDYwNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThlY2M5MjYxNmE2MjRkMjQxMTgxNGQ5ZTc4ZTUxMWJlMDRjMjg1Mzk0MzFkMzNmNjZjMWU0MDMzZDE0NGEzYzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.9a-DETpicUQVAl4h9urHnDEaiZNZg7eQIjWmteuCidY)
> > Borders and widths of `Default` and `With quick options` are different. The month range has no problem, please take a look.
> 
> Something wrong with the panelKls, I have fixed it.

![image](https://github.com/user-attachments/assets/f6a8c7e0-cb84-433e-961f-d79f3ed432ac)
![image](https://github.com/user-attachments/assets/74762029-e0f2-45dd-af7f-123916c43d13)

tolking commented on packages/components/date-picker/src/date-picker-com/basic-year-table.vue:
```diff
@@ -59,54 +67,177 @@ const startYear = computed(() => {
+    const row = (target.parentNode as HTMLTableRowElement).rowIndex
+
+    const selectedYear = row * 4 + column + startYear.value
+    const newDate = dayjs().year(selectedYear)
+    if (!props.rangeState.selecting) {
+      emit('pick', { minDate: newDate, maxDate: null })
+      emit('select', true)
     } else {
-      emit('pick', Number(year))
+      if (props.minDate && newDate >= props.minDate) {
+        emit('pick', { minDate: props.minDate, maxDate: newDate })
+      } else {
+        emit('pick', { minDate: newDate, maxDate: props.minDate })
+      }
+      emit('select', false)
     }
+  } else if (props.selectionMode === 'years') {
+    const year = target.textContent || target.innerText
+    if (event.type === 'keydown') {
+      emit('pick', castArray(props.parsedValue), false)
+      return
+    }
+    const newValue = hasClass(target, 'current')
+      ? castArray(props.parsedValue).filter((d) => d?.year() !== Number(year))
+      : castArray(props.parsedValue).concat([dayjs(year)])
+    emit('pick', newValue)
+  } else {
+    const year = target.textContent || target.innerText
```
The `selectedYear (L188)` and `year (L202/L212)` should be the same value. It is recommended to keep `selectedYear` and optimize the code.

upthen committed:
refactor(components): [date-picker] optimize YearTableClick func

btea reviewed (approved):
Thanks!
